Update changelog to document missing entries since last release (#486)
extendr/extendr: CHANGELOG.md

# Changelog

## 0.4.0

### Added

> - Support for `#[default = "..."]` in struct methods [[#436]](https://github.com/extendr/extendr/pull/436)

Reviewer: Should we explain that this applies to the parameters of methods? This way someone may think it is something applied to the method itself. Also, it does not really explains why we add this.

Author: Probably yes. Feel free to provide a better explanation (preferably as a suggestion)

- `Nullable<T>::map()` that acts on not null value and propagates `NULL` [[#446]](https://github.com/extendr/extendr/pull/446)
- [**ndarray**] Conversion from owned arrays (ie `ndarray::Array`) into `Robj` [[#450]](https://github.com/extendr/extendr/pull/450)
- [**ndarray**] Documentation for the `robj_ndarray` module [[#450]](https://github.com/extendr/extendr/pull/450)
- `Sum` for scalars like `Rint`, `Rfloat` and `Rcplx`, which accept `Iterator<Item = &Rtype>` [[#454]](https://github.com/extendr/extendr/pull/454)
- Add a new `CollectRMatrix` trait that can be used to collect arbitrary iterables into an R matrix
 [[#466]](https://github.com/extendr/extendr/pull/466)
- Documentation for `RArray::new_matrix()` [[#466]](https://github.com/extendr/extendr/pull/466)

### Changed

- [**Documentation**] Use bindgen on docs.rs, to ensure newer R features will still be documented [[#426]](https://github.com/extendr/extendr/pull/426)
- Unified how regular structs and ExternalPtr are treated (TODO: more detail needed) [[#433]](https://github.com/extendr/extendr/pull/433)

> - Updated `roxygen2` and `rextendr` versions used in integration tests [[#447]](https://github.com/extendr/extendr/pull/447)

Reviewer 2: I think this integration test is mostly for CI. Do you run this on local often? (confession: I never, except for when the CI fails)

```suggestion
```

Author: No, I don't. Yes it's probably internal.

- Bump the Rust edition from 2018 to 2021 [[#458]](https://github.com/extendr/extendr/pull/458)

> - Use `Rf_mkCharLenCE()` instead of `Rf_mkCharLen()` when creating R strings (TODO: more user friendly) [[#467]](https://github.com/extendr/extendr/pull/467)

Reviewer 2: (Honestly, I think the most user-friendly version of this is not writing at all).

```suggestion
- When converted to `STRSXP`, strings are now correctly marked as UTF-8 even on non-UTF-8 platforms (i.e., R < 4.2 on Windows), which shouldn't matter for most of the users [[#467]](https://github.com/extendr/extendr/pull/467)
```

Author: But it might impact a package built using extendr, on Windows, R < 4.2 right? So it can't be a purely internal change.

### Fixed

- The R CMD check note "Found non-API calls to R" by moving `use extendr_engine;` inside `test!` macro [[#424]](https://github.com/extendr/extendr/pull/424)

> - The Rust compiler warning "call `drop(from_raw(ptr))` if you intend to drop the `Box`" [[#434]](https://github.com/extendr/extendr/pull/434)
> - The clippy lint "this public function might dereference a raw pointer but is not marked `unsafe`" [[#451]](https://github.com/extendr/extendr/pull/451)
> - The clippy lint `clippy::extra_unused_lifetimes` [[#459]](https://github.com/extendr/extendr/pull/459)
> - The clippy lint `clippy::bool_to_int_with_if` [[#459]](https://github.com/extendr/extendr/pull/459)
> - The clippy lint `clippy::map_flatten` [[#459]](https://github.com/extendr/extendr/pull/459)
> - The clippy lint `clippy::iter_kv_map` [[#459]](https://github.com/extendr/extendr/pull/459)
> - The clippy lint `clippy::borrow_deref_ref` [[#459]](https://github.com/extendr/extendr/pull/459)

Reviewer 2: I think we don't need to list the changes mostly to silence the compiler and clippy.

```suggestion
```

Author: So you consider that Rust tooling warnings count as internal?

Reviewer 2: Yes. Let me try to be more precise:

* https://github.com/extendr/extendr/pull/434 doesn't change anything. Without the explicit `drop()`, the result of `Box::from_raw()` is dropped when the scope ends, iiuc.
* https://github.com/extendr/extendr/pull/459 doesn't change anything. We just added an annotation to silence clippy.
* https://github.com/extendr/extendr/pull/459 actually changed some logics, but it's just a kind of shortcutting.

I don't mean all the fix to clippy warnings are internal, but, this time, I think they all are.

Reviewer 2: > I ask only because consumers of extendr will see these lints and warnings, so it's not purely internal.

Have you ever seen such items appear on the changelog of a Rust crate? Or, have you ever found any of these any useful?

Author: I get that these are "cosmetic" changes that don't affect functionality. I'm saying that users themselves will encounter these warnings when they lint their own project, won't they? So it is ultimately user facing?

Reviewer 2: Okay, sorry, I was a bit confused. #451 is actually a user-facing change in that the user would see the warning on `#[extendr]` macro if they use clippy. So, I agree with you here. However, I still disagree on #434 and #459.

Author: Is that because you think they are not user facing, or because you think they are not important?

Reviewer 2: I believe #459 is not user-facing in any terms. A user doesn't usually run clippy on the dependencies. Do you?

#434 is not user-facing except when the user specifies `--verbose` option. The default `Makevars` and `Makevars.win` rextendr package provides doesn't specify it, so I bet almost no user sees the warning. I checked with the fresh `rextendr::use_extendr()`:

<details>

```
> rextendr::document()
✔ Saving changes in the open files.
ℹ Generating extendr wrapper functions for package: file581c58e24187.
ℹ Re-compiling file581c58e24187 (debug build)
── R CMD INSTALL ─────────────────────────────────────────────────────────────────────────────────────────────────────────
─  installing *source* package 'file581c58e24187' ...
   ** using staged installation
   ** libs
   using C compiler: 'gcc.exe (GCC) 12.2.0'
   rm -Rf file581c58e24187.dll ./rust/target/x86_64-pc-windows-gnu/release/libfile581c58e24187.a entrypoint.o
   gcc  -I"C:/PROGRA~1/R/R-devel/include" -DNDEBUG     -I"C:/rtools43/x86_64-w64-mingw32.static.posix/include"     -O2 -Wall -gdwarf-2 -mfpmath=sse -msse2 -mstackrealign  -UNDEBUG -Wall -pedantic -g -O0 -fdiagnostics-color=always -c entrypoint.c -o entrypoint.o
   mkdir -p ./rust/target/libgcc_mock
   # `rustc` adds `-lgcc_eh` flags to the compiler, but Rtools' GCC doesn't have
   # `libgcc_eh` due to the compilation settings. So, in order to please the
   # compiler, we need to add empty `libgcc_eh` to the library search paths.
   #
   # For more details, please refer to
   # https://github.com/r-windows/rtools-packages/blob/2407b23f1e0925bbb20a4162c963600105236318/mingw-w64-gcc/PKGBUILD#L313-L316
   touch ./rust/target/libgcc_mock/libgcc_eh.a
   # CARGO_LINKER is provided in Makevars.ucrt for R >= 4.2
   export CARGO_TARGET_X86_64_PC_WINDOWS_GNU_LINKER="x86_64-w64-mingw32.static.posix-gcc.exe" && \
   	export LIBRARY_PATH="${LIBRARY_PATH};/c/Users/Yutani/AppData/Local/Temp/Rtmpi2EFCU/file581c58e24187/src/./rust/target/libgcc_mock" && \
   	cargo build --target=x86_64-pc-windows-gnu --lib --release --manifest-path=./rust/Cargo.toml --target-dir ./rust/target
       Updating crates.io index
    Downloading crates ...
     Downloaded syn v1.0.109
      Compiling winapi v0.3.9
      Compiling proc-macro2 v1.0.51
      Compiling unicode-ident v1.0.6
      Compiling quote v1.0.23
      Compiling syn v1.0.109
      Compiling extendr-engine v0.3.1
      Compiling paste v1.0.11
      Compiling extendr-api v0.3.1
      Compiling lazy_static v1.4.0
      Compiling libR-sys v0.3.0
      Compiling extendr-macros v0.3.1
      Compiling file581c58e24187 v0.1.0 (C:\Users\Yutani\AppData\Local\Temp\Rtmpi2EFCU\file581c58e24187\src\rust)
       Finished release [optimized] target(s) in 9.04s
   gcc -shared -static-libgcc -o file581c58e24187.dll file581c58e24187-win.def entrypoint.o -L./rust/target/x86_64-pc-windows-gnu/release -lfile581c58e24187 -lws2_32 -ladvapi32 -luserenv -lbcrypt -LC:/rtools43/x86_64-w64-mingw32.static.posix/lib/x64 -LC:/rtools43/x86_64-w64-mingw32.static.posix/lib -LC:/PROGRA~1/R/R-devel/bin/x64 -lR
   installing to C:/Users/Yutani/AppData/Local/Temp/Rtmp4SMnKS/devtools_install_12d021e87779/00LOCK-file581c58e24187/00new/file581c58e24187/libs/x64
─  DONE (file581c58e24187)
✔ Writing 'R/extendr-wrappers.R'
ℹ Updating file581c58e24187 documentation
ℹ Loading file581c58e24187
Writing NAMESPACE
Writing NAMESPACE
Writing hello_world.Rd
```

</details>

Author: Thanks. This is the info I wanted. I don't really run clippy myself.

Reviewer 2: I see. It's super easy if you use VS Code.

![image](https://user-images.githubusercontent.com/1978793/221210862-7d92655b-7966-4ea6-a735-6c4db6b3a912.png)
